File: P03_WumpusWorld/P03_WumpusWorld/client/dpll.go
// dpll.go
//
// DPLL (Davis–Putnam–Logemann–Loveland) satisfiability solver used as the
// inference engine for the Wumpus World agent. The solver operates on a
// formula in Conjunctive Normal Form (CNF) and is used by the knowledge
// base (kb.go) to answer entailment queries of the form KB ⊨ φ via proof
// by refutation: KB ⊨ φ  iff  KB ∧ ¬φ is UNSAT.
package main

// Literal encodes a propositional literal as a signed integer.
//   +v  → variable v is asserted TRUE
//   -v  → variable v is asserted FALSE
// Variable identifiers must be strictly positive.
type Literal int

// Var returns the underlying variable identifier (always positive).
func (l Literal) Var() int {
	if l < 0 {
		return int(-l)
	}
	return int(l)
}

// IsPos reports whether the literal is a positive (un-negated) variable.
func (l Literal) IsPos() bool { return l > 0 }

// Clause is a disjunction (logical OR) of literals.
type Clause []Literal

// CNF is a conjunction (logical AND) of clauses.
type CNF []Clause

// simplify evaluates every clause against the current partial assignment:
//   - if the clause already contains a literal satisfied by the assignment,
//     the clause is removed (it is already true);
//   - any literal falsified by the assignment is dropped;
//   - a clause reduced to the empty clause signals a contradiction.
//
// It returns the simplified CNF and a boolean flag that is false when an
// empty clause appears (i.e. the formula became UNSAT under this partial
// assignment).
func simplify(clauses CNF, assignment map[int]bool) (CNF, bool) {
	result := make(CNF, 0, len(clauses))
	for _, clause := range clauses {
		satisfied := false
		remaining := make(Clause, 0, len(clause))
		for _, lit := range clause {
			v := lit.Var()
			if val, assigned := assignment[v]; assigned {
				if lit.IsPos() == val {
					satisfied = true
					break
				}
				// literal is false under the assignment – drop it
			} else {
				remaining = append(remaining, lit)
			}
		}
		if satisfied {
			continue
		}
		if len(remaining) == 0 {
			return nil, false // empty clause → UNSAT
		}
		result = append(result, remaining)
	}
	return result, true
}

// chooseBranchVar returns the first unassigned variable appearing in the
// clauses, or 0 if no such variable exists.
func chooseBranchVar(clauses CNF, assignment map[int]bool) int {
	for _, clause := range clauses {
		for _, lit := range clause {
			v := lit.Var()
			if _, assigned := assignment[v]; !assigned {
				return v
			}
		}
	}
	return 0
}

// copyAssignment returns a shallow copy of the assignment map.
func copyAssignment(a map[int]bool) map[int]bool {
	c := make(map[int]bool, len(a))
	for k, v := range a {
		c[k] = v
	}
	return c
}

// dpll is the recursive core of the DPLL algorithm. It returns true if the
// given CNF is satisfiable under an extension of the provided partial
// assignment. The algorithm applies unit propagation, then branches on an
// unassigned variable (trying true first, then false).
func dpll(clauses CNF, assignment map[int]bool) bool {
	// Base case: all clauses satisfied.
	if len(clauses) == 0 {
		return true
	}

	// Unit propagation loop. Whenever a clause has a single literal, that
	// literal must be true; assign it and re-simplify.
	for {
		unitFound := false
		for _, clause := range clauses {
			if len(clause) != 1 {
				continue
			}
			lit := clause[0]
			v := lit.Var()
			val := lit.IsPos()
			if cur, exists := assignment[v]; exists {
				if cur != val {
					return false // conflict between two unit clauses
				}
				continue
			}
			assignment[v] = val
			var ok bool
			clauses, ok = simplify(clauses, map[int]bool{v: val})
			if !ok {
				return false
			}
			if len(clauses) == 0 {
				return true
			}
			unitFound = true
			break // restart the scan with the simplified clauses
		}
		if !unitFound {
			break
		}
	}

	// Branching heuristic: pick the first unassigned variable.
	v := chooseBranchVar(clauses, assignment)
	if v == 0 {
		return true // every remaining clause was satisfied
	}

	// Try v = true.
	asgnTrue := copyAssignment(assignment)
	asgnTrue[v] = true
	if sub, ok := simplify(clauses, map[int]bool{v: true}); ok && dpll(sub, asgnTrue) {
		return true
	}

	// Try v = false.
	asgnFalse := copyAssignment(assignment)
	asgnFalse[v] = false
	sub, ok := simplify(clauses, map[int]bool{v: false})
	return ok && dpll(sub, asgnFalse)
}

// IsSAT reports whether the given CNF is satisfiable.
func IsSAT(clauses CNF) bool {
	cp := make(CNF, len(clauses))
	copy(cp, clauses)
	return dpll(cp, make(map[int]bool))
}

// Entails reports whether the knowledge base entails the single literal
// query, implemented as proof by refutation:
//
//	KB ⊨ query  ⇔  KB ∧ ¬query  is UNSAT.
func Entails(kb CNF, query Literal) bool {
	negated := Literal(-query) // flip the sign of the query literal
	test := make(CNF, len(kb)+1)
	copy(test, kb)
	test[len(kb)] = Clause{negated}
	return !dpll(test, make(map[int]bool))
}

File: P03_WumpusWorld/P03_WumpusWorld/python_client/dpll.py
"""
dpll.py
=======
DPLL (Davis–Putnam–Logemann–Loveland) satisfiability solver used as the
inference engine for the Wumpus World agent.

The solver operates on a formula in Conjunctive Normal Form (CNF) and is
used by the knowledge base (kb.py) to answer entailment queries of the
form KB ⊨ φ via proof by refutation:

    KB ⊨ φ  iff  KB ∧ ¬φ  is UNSAT.

A literal is represented as a non-zero integer:
  +v  → variable v is TRUE
  -v  → variable v is FALSE
Variable identifiers must be strictly positive integers.

A Clause  is a list of literals  (disjunction / OR).
A CNF     is a list of Clauses   (conjunction / AND).
"""

from __future__ import annotations
from typing import Dict, List, Optional, Tuple

# Type aliases
Literal = int          # non-zero signed integer
Clause  = List[Literal]
CNF     = List[Clause]


def _var(lit: Literal) -> int:
    """Return the underlying positive variable id."""
    return abs(lit)


def _is_pos(lit: Literal) -> bool:
    """Return True if the literal asserts the variable TRUE."""
    return lit > 0


# ---------------------------------------------------------------------------
# Core helpers
# ---------------------------------------------------------------------------

def _simplify(clauses: CNF, assignment: Dict[int, bool]) -> Tuple[Optional[CNF], bool]:
    """
    Evaluate every clause against the current partial assignment.

    - Clauses that contain a satisfied literal are removed.
    - Falsified literals are dropped from their clause.
    - A clause reduced to the empty set signals UNSAT (returns None, False).

    Returns (simplified_cnf, True) or (None, False) on contradiction.
    """
    result: CNF = []
    for clause in clauses:
        satisfied = False
        remaining: Clause = []
        for lit in clause:
            v = _var(lit)
            if v in assignment:
                if _is_pos(lit) == assignment[v]:
                    satisfied = True
                    break
                # literal is false under assignment – drop it
            else:
                remaining.append(lit)
        if satisfied:
            continue
        if not remaining:
            return None, False   # empty clause → UNSAT
        result.append(remaining)
    return result, True


def _choose_branch_var(clauses: CNF, assignment: Dict[int, bool]) -> int:
    """Return the first unassigned variable found in the clauses, or 0."""
    for clause in clauses:
        for lit in clause:
            v = _var(lit)
            if v not in assignment:
                return v
    return 0


def _dpll(clauses: CNF, assignment: Dict[int, bool]) -> bool:
    """
    Recursive DPLL core.  Returns True iff the CNF is satisfiable under
    some extension of the supplied partial assignment.

    Applies unit propagation first, then branches on an unassigned variable
    (trying True first, then False).
    """
    # Unit-propagation loop
    while True:
        unit_found = False
        for clause in clauses:
            if len(clause) != 1:
                continue
            lit = clause[0]
            v   = _var(lit)
            val = _is_pos(lit)

            if v in assignment:
                if assignment[v] != val:
                    return False    # conflict between two unit clauses
                continue

            assignment[v] = val
            clauses, ok = _simplify(clauses, {v: val})
            if not ok:
                return False
            if not clauses:
                return True
            unit_found = True
            break                   # restart scan with simplified clauses

        if not unit_found:
            break

    # All clauses satisfied?
    if not clauses:
        return True

    # Choose a branching variable
    v = _choose_branch_var(clauses, assignment)
    if v == 0:
        return True   # every remaining clause was already satisfied

    # Branch: try v = True
    asgn_true = dict(assignment)
    asgn_true[v] = True
    sub, ok = _simplify(clauses, {v: True})
    if ok and _dpll(sub, asgn_true):
        return True

    # Branch: try v = False
    asgn_false = dict(assignment)
    asgn_false[v] = False
    sub, ok = _simplify(clauses, {v: False})
    return ok and _dpll(sub, asgn_false)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def is_sat(clauses: CNF) -> bool:
    """Return True if the CNF formula is satisfiable."""
    # Work on a copy so the caller's formula is not modified.
    return _dpll([list(c) for c in clauses], {})


def entails(kb: CNF, query: Literal) -> bool:
    """
    Return True when the knowledge base entails `query`, using proof by
    refutation:

        KB ⊨ query  ⇔  KB ∧ ¬query  is UNSAT.
    """
    negated = -query                           # flip the sign
    test = [list(c) for c in kb] + [[negated]]
    return not _dpll(test, {})
